File: ADS_2_POO/Python Version/tituloEleitor.py
from documentos import Documentos
from documentosFormatacao import DocumentosFormatacao

class TituloEleitor(Documentos):
    def __init__(self, numero_input, tamanho_com_digitos, tamanho_sem_digitos):
        super().__init__(numero_input, tamanho_com_digitos, tamanho_sem_digitos)
        self.peso_um = [2, 3, 4, 5, 6, 7, 8, 9]
        self.peso_final = [7, 8, 9]
        self.numero = DocumentosFormatacao.formata_numero(numero_input, tamanho_com_digitos, tamanho_sem_digitos)
        self.numero_correto = (self.numero[:tamanho_sem_digitos] + self.calcular_digito(self.numero, self.peso_um))
        self.numero_correto = self.numero_correto + self.calcular_digito(
        self.numero_correto[self.tamanho_sem_digitos - 2:3], self.peso_final)
        self.mascara_numero = f"{self.numero_correto[:4]} {self.numero_correto[4:tamanho_sem_digitos - 2]} {self.numero_correto[tamanho_sem_digitos - 2:][:2]} {self.numero_correto[tamanho_com_digitos - 2:][:2]}"

    def calcular_digito(self, numero_input, pesos):
        soma = sum(int(numero_input[i]) * pesos[i] for i in range(min(len(numero_input), len(pesos))))
        resto = soma % 11

        digitos_dos_estados = numero_input[-2:]

        if digitos_dos_estados == "01" or digitos_dos_estados == "02":
            if resto == 0:
                return "1"  # Condição especial para SP e MG

        return '0' if resto > 9 else str(resto)

    def validar(self):
        resposta = f"{self.mascara_numero} {self.numero_correto} "

        if len(self.numero) == self.tamanho_com_digitos - 2:
            self.numero = self.numero + self.calcular_digito(self.numero, self.peso_um)

        if len(self.numero) == self.tamanho_com_digitos - 1:
            self.numero = self.numero + self.calcular_digito(self.numero[self.tamanho_sem_digitos - 2:3], self.peso_final)

        resposta = resposta + ("Valido" if self.numero_correto == self.numero else "Invalido")
        return resposta

File: ADS_2_POO/ADS2/TituloEleitor.cs

public class TituloEleitor : Documentos
{

    public TituloEleitor(string numeroInput, int tamanhoComDigitos, int tamanhoSemDigitos) : base(numeroInput, tamanhoComDigitos, tamanhoSemDigitos)
    {
        pesoUm = new int[] { 2, 3, 4, 5, 6, 7, 8, 9 };
        pesoFinal = new int[] { 7, 8, 9 };
        numero = DocumentosFormatacao.FormataNumero(numero, tamanhoComDigitos, tamanhoSemDigitos);
        numeroCorreto = (numero[..tamanhoSemDigitos] + CalcularDigito(numero, pesoUm));
        numeroCorreto = numeroCorreto + CalcularDigito(numeroCorreto.Substring(tamanhoSemDigitos - 2, 3), pesoFinal); 

        mascaraNumero = numeroCorreto.Substring(0, 4) + " " + numeroCorreto.Substring(4, tamanhoSemDigitos - 6) + " " + numeroCorreto.Substring(tamanhoSemDigitos - 2, 2) + " " + numeroCorreto.Substring(tamanhoComDigitos - 2, 2);

    }

    public override string CalcularDigito(string numeroInput, int[] pesos)
    {
        int soma = 0;
        int length = Math.Min(numeroInput.Length, pesos.Length);//Por que pesos.Lenght - 1 n funcionou?

        for (int i = 0; i < length; i++)
        {
            soma += int.Parse(numeroInput[i].ToString()) * pesos[i];
        }

        int resto = soma % 11;

        resto = (resto > 9) ? 0 : resto;
        string digitosDosEstados = numero.Substring(tamanhoSemDigitos - 2, 2);

        if (digitosDosEstados == "01" || digitosDosEstados == "02")
        {
            if (resto == 0)
                return "1"; // Condição especial para SP e MG
        }
        
        return resto.ToString(); 
        
    }

    public override string Validar()
    {
        string resposta = mascaraNumero + " " + numeroCorreto + " ";
        if (numero.Length == tamanhoComDigitos - 2)
        {
            numero = (numero + CalcularDigito(numero, pesoUm));
        }

        if (numero.Length == tamanhoComDigitos - 1)
        {
            numero = numero + CalcularDigito(numero.Substring(tamanhoSemDigitos - 2, 3), pesoFinal);
        }

        resposta = resposta + ((numeroCorreto == numero) ? "Valido" : "Invalido");
        return resposta;
    }
}

